kolibri/auth/test/test_permissions.py

```diff
@@ -1,454 +1,568 @@
-#         self.assertFalse(self.coach1.has_perm(self.AUTH_REMOVE_LEARNER, self.learner_groups[1]))
-
-#     def test_remove_learner_specific_for_learner(self):
-#         self.assertFalse(self.learner1.has_perm(self.AUTH_REMOVE_LEARNER, self.learner_groups[0]))
-
-#     def test_remove_learner_rejects_non_classroom_objects(self):
-#         with self.assertRaises(InvalidPermission):
-#             self.admin.has_perm(self.AUTH_REMOVE_LEARNER, obj={})
-
-#     # noqa ##################################
-#     # noqa #                               ##
-#     # noqa #   auth.add_facility_admin     ##
-#     # noqa #                               ##
-#     # noqa ##################################
-#     AUTH_ADD_FACILITY_ADMIN = 'auth.add_facility_admin'
-
-#     def test_add_facility_admin_for_admin(self):
-#         self.assertTrue(self.admin.has_perm(self.AUTH_ADD_FACILITY_ADMIN))
-
-#     def test_add_facility_admin_for_coach(self):
-#         self.assertFalse(self.coach2.has_perm(self.AUTH_ADD_FACILITY_ADMIN))
-
-#     def test_add_facility_admin_for_learner(self):
-#         self.assertFalse(self.learner1.has_perm(self.AUTH_ADD_FACILITY_ADMIN))
-
-#     # noqa ##################################
-#     # noqa #                               ##
-#     # noqa #   auth.remove_facility_admin  ##
-#     # noqa #                               ##
-#     # noqa ##################################
-#     AUTH_REMOVE_FACILITY_ADMIN = 'auth.remove_facility_admin'
-
-#     def test_remove_facility_admin_for_admin(self):
-#         self.assertTrue(self.admin.has_perm(self.AUTH_REMOVE_FACILITY_ADMIN))
-
-#     def test_remove_facility_admin_for_coach(self):
-#         self.assertFalse(self.coach2.has_perm(self.AUTH_REMOVE_FACILITY_ADMIN))
-
-#     def test_remove_facility_admin_for_learner(self):
-#         self.assertFalse(self.learner1.has_perm(self.AUTH_REMOVE_FACILITY_ADMIN))
+"""
+Tests of the permissions on specific models in the auth app. For tests of the permissions system itself, see test_permission_classes.py
+"""
+
+from __future__ import absolute_import, print_function, unicode_literals
+
+from django.test import TestCase
+
+from .helpers import create_dummy_facility_data
+
+from ..constants import role_kinds
+from ..models import DeviceOwner, Facility, Classroom, LearnerGroup, Role, FacilityUser
+
+
+class FacilityPermissionsTestCase(TestCase):
+    """
+    Tests of permissions for reading/modifying Facility instances
+    """
+
+    def setUp(self):
+        self.data1 = create_dummy_facility_data()
+        self.data2 = create_dummy_facility_data()
+        self.device_owner = DeviceOwner.objects.create(username="boss")
+
+    def test_facility_users_cannot_create_facility(self):
+        """ FacilityUsers can't create new Facilities, regardless of their roles """
+        unsaved_facility = Facility(name="Home")
+        self.assertFalse(unsaved_facility.user_can_create(self.data1["facilityadmin"]))
```

Comment:
> create here is not a row-level permission, right? In contrast to update or read, it applies only to whole Model classes and not rows, if I'm getting it correctly

For our use cases, `create` permission checks *do* typically need to be "row-level" (in the sense that whether or not creating an instance should be allowed depends on the specific contents of the provided data). For example, we allow an admin for a facility to create a `FacilityUser` [for their own `Facility`](https://github.com/learningequality/kolibri/pull/51/files#diff-51e21fa50df52b4e8505f7ae33dfad7eR256), but we don't let them create one [for another `Facility`](https://github.com/learningequality/kolibri/pull/51/files#diff-51e21fa50df52b4e8505f7ae33dfad7eR267). Similarly, an admin for a `Classroom` can create an admin `Role` object [for that `Classroom`](https://github.com/learningequality/kolibri/pull/51/files#diff-51e21fa50df52b4e8505f7ae33dfad7eR467), but cannot create one [for a different `Classroom`](https://github.com/learningequality/kolibri/pull/51/files#diff-51e21fa50df52b4e8505f7ae33dfad7eR469) (for which she is not an admin).

> Incidentally do you have an example of row-level permissions with DRF you can link to, for my own edification?

High-level description is here: http://www.django-rest-framework.org/api-guide/permissions/#object-level-permissions

And the second example here shows a simple case: http://www.django-rest-framework.org/api-guide/permissions/#examples

However, thanks for getting me to look into this, as it seems DRF doesn't behave as I had expected it to (from the documentation) in the case of creation (`POST`). Instead, it seems to be recommended practice to [use something like `pre_save`](https://github.com/kevinastone/django-api-rest-and-angular/blob/00888c990a831f7adaba6ace5fd800b3fe7140e1/example/api/api.py#L30) or validation methods to ensure the data being saved is appropriate. Others seemed to be under the same impression as me, and hence implemented things [like this](http://stackoverflow.com/a/18646798), which won't actually be checked in the "create" case.

> if an instance is required as an implementation detail we can hide that from the dev

I agree that it could be done with passing the data as a dictionary, rather than a model instance, and then we could instantiate a model instance internally as needed (and it will be needed, for running `ensure_dataset`, etc, prior to permission-checking). I'm pondering this along with the suggestions around method naming referenced below, and will come back on that soon.

> For example if there's only a 1% chance that someone will write an error but we write 10000 LOCs, then we'll still expect to see a lot of errors. Can we eliminate or diminish the possibility of an error?

I don't think this decreases the chances of errors all that much; the main source of errors would be devs not calling the permissions checks in the first place (whether for reading, creating, updating, etc). We'll just need to police that, when data is being manipulated within Django views, using code reviews and tests. Data accessed through the HTTP API, on the other hand, will all automatically pass through the permission checks.

More soon.